Feature/opcodes/splice

akifuj opened the pull request:
close #63 
close #67 
close #340 

I implemented `OP_CAT` and `OP_SPLIT`

I could not implement `OP_BIN2NUM` and `OP_NUM2BIN` and just created the files for them. 
If anyone can implement it, please help.
FYI, `MinimallyEncode`, the function used in them, is defined in [script.cpp](
https://github.com/Bitcoin-ABC/bitcoin-abc/blob/master/src/script/script.cpp). Refer to it.

Besides, I just added files of `OP_SUBSTR`, `OP_LEFT`, and `OP_RIGHT` without implementation for `mainProcess` as well, for BitcoinABC does not include these disabled Opcodes.  I think it might be better to delete them from `Opcode` like BitcoinABC. Tell me your opinions.

usatie commented on BitcoinCashKit/Core/Scripts/OP_CODE/Splice/OP_SIZE.swift:
```diff
@@ -0,0 +1,40 @@
+        try context.assertStackHeightGreaterThanOrEqual(1)
+        let input: Data = context.data(at: -1)
+        let count: Int = input.count
+        try context.pushToStack(Int32(count))
```
`try context.pushToStack(Int32(input.count))`
is enough.

usatie replied:
And also the name `input` is not good.
I think it should be `x` or something like that.

usatie commented on BitcoinCashKit/Core/Scripts/OP_CODE/Splice/OP_SPLIT.swift:
```diff
@@ -0,0 +1,51 @@
+        
+        // Make sure the split point is apropriate.
+        let position: Int32 = try context.number(at: -1)
+        guard position <= MemoryLayout.size(ofValue: data) else {
```
`guard position <= data.count` is shorter.

usatie replied:
I think `MemoryLayout.size(ofValue: )` should be used for non-Data class.

usatie commented on BitcoinCashKit/Core/Scripts/OP_CODE/Splice/OP_SPLIT.swift:
```diff
@@ -0,0 +1,51 @@
+
+        // Replace existing stack values by the new values.
+        context.stack[data.count - 2] = n1
+        context.stack[data.count - 1] = n2
```
It has to be this,
```
context.stack[context.stack.count - 2] = n1
context.stack[context.stack.count - 1] = n2
```

akifuj commented:
- Added new issues for `OP_BIN2NUM` and `OP_NUM2BIN`
- Removed files of `OP_LEFT`, `OP_RIGHT`, and `OP_SUBSTR`.
- Added unit tests for `OP_CAT`, `OP_SIZE`, and `OP_SPLIT`

usatie commented on BitcoinCashKit/Core/Scripts/OP_CODE/Splice/OP_BIN2NUM.swift:
```diff
@@ -28,12 +28,10 @@ import Foundation
-        try context.assertStackHeightGreaterThanOrEqual(1)
-        
-        // TODO: not implemented
+        // not implemented
```
Please delete `mainProcess()` so that notImplementedError is thrown.

usatie commented on BitcoinCashKit/Core/Scripts/OP_CODE/Splice/OP_NUM2BIN.swift:
```diff
@@ -28,19 +28,10 @@ import Foundation
-        context.stack.removeLast()
-        
-        // TODO: not implemented
+        // not implemented
```
Please delete mainProcess() so that notImplementedError is thrown when it's called.
